pkg/resource/deploy/plan.go

```diff
@@ -0,0 +1,612 @@
+	InputDiff PlanDiff
+	// the resource's output properties we expect to change (only set for RegisterResourceOutputs)
+	OutputDiff *PlanDiff
+	// an optional parent URN for this resource.
+	Parent resource.URN
+	// true to protect this resource from deletion.
+	Protect bool
+	// dependencies of this resource object.
+	Dependencies []resource.URN
+	// the provider to use for this resource.
+	Provider string
+	// the set of dependencies that affect each property.
+	PropertyDependencies map[resource.PropertyKey][]resource.URN
+	// true if this resource should be deleted prior to replacement.
+	DeleteBeforeReplace *bool
+	// a list of property names to ignore during changes.
+	IgnoreChanges []string
+	// outputs that should always be treated as secrets.
+	AdditionalSecretOutputs []resource.PropertyKey
+	// additional URNs that should be aliased to this resource.
+	Aliases []resource.URN
+	// the expected ID of the resource, if any.
+	ID resource.ID
+	// an optional config object for resource options
+	CustomTimeouts resource.CustomTimeouts
+}
+
+func NewPlanDiff(inputDiff *resource.ObjectDiff) PlanDiff {
+	var adds resource.PropertyMap
+	var deletes []resource.PropertyKey
+	var updates resource.PropertyMap
+
+	var diff PlanDiff
+	if inputDiff != nil {
+		adds = inputDiff.Adds
+		updates = make(resource.PropertyMap)
+		for k := range inputDiff.Updates {
+			updates[k] = inputDiff.Updates[k].New
+		}
+		deletes = make([]resource.PropertyKey, len(inputDiff.Deletes))
+		i := 0
+		for k := range inputDiff.Deletes {
+			deletes[i] = k
+			i = i + 1
+		}
+
+		diff = PlanDiff{Adds: adds, Deletes: deletes, Updates: updates}
+	}
+
+	return diff
+}
+
+func NewGoalPlan(checkedInputs resource.PropertyMap, inputDiff *resource.ObjectDiff, goal *resource.Goal) *GoalPlan {
+	if goal == nil {
+		return nil
+	}
+
+	var diff = NewPlanDiff(inputDiff)
+
+	return &GoalPlan{
+		Type:                    goal.Type,
+		Name:                    goal.Name,
+		Custom:                  goal.Custom,
+		CheckedInputs:           checkedInputs,
+		InputDiff:               diff,
+		OutputDiff:              nil,
+		Parent:                  goal.Parent,
+		Protect:                 goal.Protect,
+		Dependencies:            goal.Dependencies,
+		Provider:                goal.Provider,
+		PropertyDependencies:    goal.PropertyDependencies,
+		DeleteBeforeReplace:     goal.DeleteBeforeReplace,
+		IgnoreChanges:           goal.IgnoreChanges,
+		AdditionalSecretOutputs: goal.AdditionalSecretOutputs,
+		Aliases:                 goal.Aliases,
+		ID:                      goal.ID,
+		CustomTimeouts:          goal.CustomTimeouts,
+	}
+}
+
+// A ResourcePlan represents the planned goal state and resource operations for a single resource. The operations are
+// ordered.
+type ResourcePlan struct {
+	Goal    *GoalPlan
+	Ops     []StepOp
+	Outputs resource.PropertyMap
+}
+
+func (rp *ResourcePlan) diffURNs(a, b []resource.URN) (message string, changed bool) {
```

Comment:
Are these sorted/deduped, that is are these `[]` slices actually representing sets? Should we be defensive and sort/dedup here?

Reply:
Yes they are sets that's why we make the two string maps and then diff that.